simulation/utils/generate_map.py

```diff
@@ -0,0 +1,241 @@
+
+    def getPoint(self, x_offset=0):
+        point = (self.x + x_offset, self.y)
+        return point
+
+    def __getitem__(self, key):
+        return self.x if not key else self.y
+
+
+class Map:
+    def __init__(self, starting_point, x_offset, resolution=(1280, 720), seed=None):
+        self.points_before_interpolation = []
+        self.map_points = []
+        self.segments = []
+        self.x_offset = x_offset
+        self.append(starting_point)
+        self.resolution = resolution
+        self.seed = seed
+
+    # ************
+    def append(self, point):
+        self.points_before_interpolation.append(point)
+
+    def __getitem__(self, key):
+        """Get point of the map: e.g. game_map[7] ---> (x7,y7)"""
+        if not len(self.map_points):
+            return self.points_before_interpolation[key]
+        else:
+            return self.map_points[key]
+
+    def get_data_as_points(self):
+        """Get map in format [(x0,y0), (x1,y1), (x2,y2), ...]"""
+        return self.map_points
+
+    def get_data_as_segments(self):
+        """Get map in format [((x1,y1), (x1,y1)), ((x1,y1), (x2,y2)), ((x2,y2), (x3,y3)), ...]"""
+        return self.segments
+
+    def get_seed(self):
+        """Get the seed of the map"""
+        return self.seed
+
+    def X(self):
+        """Get list of x coordinates"""
+        xs = [point.x for point in self.points_before_interpolation] if not len(self.map_points) else [point[0] for point in self.map_points]
+        return xs
+
+    def Y(self):
+        """Get list of y coordinates"""
+        ys = [point.y for point in self.points_before_interpolation] if not len(self.map_points) else [point[1] for point in self.map_points]
+        return ys
+
+    def __len__(self):
+        """ Length of map (points, not segments) """
+        return len(self.points_before_interpolation) if not len(self.map_points) else len(self.map_points)
+
+    # ************
+    def __delitem__(self, index):
+        if not len(self.map_points):
+            del self.points_before_interpolation[index]
+        else:
+            del self.map_points[index]
+
+    # ************
+    def is_crossing(self, i, j):
+        """Function that checks if two lines ((X[i],Y[i]), (X[i+1], Y[i+1])), ((X[j],Y[j]), (X[j+1],Y[j+1])) intersect"""
+        if i == j:
+            return False
+        try:
+            return intersect(self.points_before_interpolation[i], self.points_before_interpolation[i + 1],
+                             self.points_before_interpolation[j], self.points_before_interpolation[j + 1])
+        except IndexError:
+            print(f'i = {i}, j = {j}, size = {len(self)}')
+
+    # ************
+    def exterminate_loops(self, step, y_res, ran=40):
+        """Function that gets rid of most of the loops in map so as to make it a little less crazy"""
+        i = 0
+        size = len(self.points_before_interpolation)
+        while i < size - 1:
```

Comment:
Całkiem sporo się dzieje w tym while'u, może da się to rozbić jakoś na mniejsze segmenty czy metody?